app/controllers/fhir/r4/api_controller.rb

```diff
@@ -426,6 +426,132 @@ def all
     status_ok(bundle) && return
   end
 
+  # Kick off async generation of bulk FHIR data for all accessible patients
+  #
+  # GET /fhir/r4/Patient/$export
+  def bulk_data_export
+    # Require all scopes for all possible resources being returned
+    return if doorkeeper_authorize!(*PATIENT_READ_SCOPES)
+    return if doorkeeper_authorize!(*OBSERVATION_READ_SCOPES)
+    return if doorkeeper_authorize!(*QUESTIONNAIRE_RESPONSE_READ_SCOPES)
+    return if doorkeeper_authorize!(*RELATED_PERSON_READ_SCOPES)
+    return if doorkeeper_authorize!(*IMMUNIZATION_READ_SCOPES)
+    return if doorkeeper_authorize!(*PROVENANCE_READ_SCOPES)
+
+    unless accept_header?
+      status_not_acceptable_with_custom_errors(["'Accept' header must have a value of 'application/fhir+json'," \
+        " or the '_format' parameter must be one of 'json', 'application/json' or 'application/fhir+json'"]) && return
+    end
+
+    unless prefer_header?
+      status_unprocessable_entity_with_custom_errors(["'Prefer' header must have value of 'respond-async'"],
+                                                     '') && return
+    end
+
+    unless @m2m_workflow
+      status_unauthorized_with_custom_errors(['Bulk export requires a client application registered for the Backend Services Workflow']) && return
+    end
+
+    if params.key?(:_type) || params.key?(:_outputFormat)
+      status_unprocessable_entity_with_custom_errors(['The _type and _outputFormat parameters are unsupported'], '') && return
+    end
+
+    begin
+      since = params.permit(:_since)[:_since]
+      since = DateTime.strptime(since, '%Y-%m-%dT%H:%M:%S%z') unless since.blank?
+    rescue Date::Error
+      status_unprocessable_entity_with_custom_errors(['Invalid Date in _since parameter. Please use the FHIR instant datatype'], '') && return
+    end
+
+    client_app = current_client_application
+    if client_app.exported_recently?
+      response['Retry-After'] = client_app.retry_bulk_export_after.httpdate
+      status_too_many_requests_with_custom_errors(
+        ["Client already initiated an export of this type in the last #{Rails.configuration.api['bulk_export_retry_after_minutes']} minutes. " \
+        'Please try again later']
+      ) && return
+    end
+
+    # Create the download to uniquely id this request, and queue the ExportFhirJob
+    download = ApiDownload.create(application_id: client_app.id, url: request.url)
+    export_job = ExportFhirJob.perform_later(client_app, download, { since: since })
+    # Add the job_id of the job to the download for reference in status requests
+    download.update(job_id: export_job.provider_job_id)
+
+    respond_to do |format|
+      format.any { head(:accepted, content_location: "#{root_url}fhir/r4/ExportStatus/#{download.id}") }
+    end
+  end
+
+  # Give an update on the status of a bulk FHIR data request
+  #
+  # GET /fhir/r4/ExportStatus/[:id]
+  def export_status
+    id = params.require(:id)
+    download = ApiDownload.find_by_id(id)
```

Comment:
This query needs to be scoped. I think you have access to client_app.id here, so use that.

Reply:
Scoped by the token id in https://github.com/SaraAlert/SaraAlert/pull/987/commits/5be75cbe252a41cd26747d9b35a057ef7a118ab7.